[question] How do you set the NDK path for Android builds in CMakeToolchain?

I'm trying to figure out how to build for Android using the Conan 2.0 `CMakeToolchain`. I get this error:

`ConanException: CMakeToolchain needs tools.android:ndk_path configuration defined`

The docs mention `tools.android:ndk_path`, but not how to set it. There are a few clues in the source code, but all the things I've tried in `conanfile.py` and `conan.conf` haven't worked. This was very easy with the `cmake` generator, I just did this:

`cmake.definitions["CMAKE_TOOLCHAIN_FILE"] = os.getenv("ANDROID_NDK_ROOT") + "/build/cmake/android.toolchain.cmake"`

...but it doesn't work anymore.

Currently, I share this Python script with my Conan 1.x packages to define build settings for desktop and mobile platform builds:
https://github.com/ssrobins/conan-cmake_utils/blob/217df94/cmake_utils.py. 

In addition to being stuck on this, I'm a bit concerned that Conan 2.0 will make it not possible or at least harder to use custom toolchains, which is critical for close-sourced platforms that don't have any built-in CMake support.

Thanks @memsharded, putting that code in `~/.conan/profiles/default` allowed NDK to be found. Unfortunately, it failed right after with:
```
Invalid Android STL: None.
CMake Error: CMake was unable to find a build program corresponding to "Ninja Multi-Config".  CMAKE_MAKE_PROGRAM is not set.  You probably need to select a different build tool.
```
I tried adding `ANDROID_STL` and `CMAKE_MAKE_PROGRAM`, even though `ninja` is in my `PATH` and is working in non-Conan 2.0 packages, but it didn't help. Here's my full `generate()` method:
```
def generate(self):
    tc = CMakeToolchain(self)
    if self.settings.os == "Android":
        tc.generator = "Ninja Multi-Config"
        tc.variables["CMAKE_MAKE_PROGRAM"] = "/usr/local/bin/ninja"
        tc.variables["ANDROID_STL"] = "c++_static"
    elif self.settings.os == "iOS":
        tc.generator = "Xcode"
        tc.variables["CMAKE_OSX_DEPLOYMENT_TARGET"] = self.settings.os.version
    elif self.settings.os == "Linux":
            tc.generator = "Ninja Multi-Config"
    elif self.settings.os == "Macos":
        tc.generator = "Xcode"
        tc.variables["CMAKE_OSX_DEPLOYMENT_TARGET"] = self.settings.os.version
    tc.generate()
    deps = CMakeDeps(self)
    deps.generate()
```
The `CMAKE_OSX_DEPLOYMENT_TARGET` for macOS and iOS are working so this seems to only affect the Android builds.

For Conan 1.x, this is all I had to add to my `conanfile.py` to build Android:
```
if settings.os == "Android":
    cmake.generator = "Ninja Multi-Config"
    if settings.arch == "armv7":
        cmake.definitions["ANDROID_ABI"] = "armeabi-v7a"
    elif settings.arch == "armv8":
        cmake.definitions["ANDROID_ABI"] = "arm64-v8a"
    cmake.definitions["ANDROID_TOOLCHAIN"] = "clang"
    cmake.definitions["ANDROID_STL"] = "c++_static"
    cmake.definitions["CMAKE_TOOLCHAIN_FILE"] = os.getenv("ANDROID_NDK_ROOT") + "/build/cmake/android.toolchain.cmake"
```
I was just able to pass CMake params exactly how it would work without Conan. Yes it's more code than if it was taken care of for me, but I appreciated the transparency and simplicity. It let me use the `ANDROID_NDK_ROOT` environment variable as the single definition for NDK on my system that works for CMake and Conan 1.x. It's also starting to become the standard in cloud CI environments like GitHub Actions.

For the Android toolchain as well as any custom toolchain, I would love for this to be done with a setting in the conanfile.py, like it was before. The `cmake` interface is excellent in Conan 1.x, I don't understand why it's being made less discoverable and spread to multiple files for Conan 2.x.

Finally got Android builds working with Conan 2.0. Got past this error:
```
Invalid Android STL: None.
CMake Error: CMake was unable to find a build program corresponding to "Ninja Multi-Config".  CMAKE_MAKE_PROGRAM is not set.  You probably need to select a different build tool.
```

I just had to set `compiler.libcxx=c++_static`. I didn't realize the Android STL setting came from there. Setting `compiler.libcxx=c++_shared` works too. I guess CMake was interpreting the early error as a problem finding the make program, hence the second error.

--- a/conan/tools/cmake/toolchain.py
+++ b/conan/tools/cmake/toolchain.py
@@ -254,16 +254,13 @@ class ParallelBlock(Block):
 
 
 class AndroidSystemBlock(Block):
-    # TODO: fPIC, fPIE
-    # TODO: RPATH, cross-compiling to Android?
-    # TODO: libcxx, only libc++ https://developer.android.com/ndk/guides/cpp-support
 
     template = textwrap.dedent("""
         # New toolchain things
-        set(ANDROID_PLATFORM {{ CMAKE_SYSTEM_VERSION }})
-        set(ANDROID_STL {{ CMAKE_ANDROID_STL_TYPE }})
-        set(ANDROID_ABI {{ CMAKE_ANDROID_ARCH_ABI }})
-        include({{ CMAKE_ANDROID_NDK }}/build/cmake/android.toolchain.cmake)
+        set(ANDROID_PLATFORM {{ ANDROID_PLATFORM }})
+        set(ANDROID_STL {{ ANDROID_STL }})
+        set(ANDROID_ABI {{ ANDROID_ABI }})
+        include({{ ANDROID_NDK_PATH }}/build/cmake/android.toolchain.cmake)
         """)
 
     def context(self):
@@ -277,19 +274,19 @@ class AndroidSystemBlock(Block):
                        "armv8": "arm64-v8a"}.get(str(self._conanfile.settings.arch))
 
         # TODO: only 'c++_shared' y 'c++_static' supported?
+        #  https://developer.android.com/ndk/guides/cpp-support
         libcxx_str = str(self._conanfile.settings.compiler.libcxx)
 
-        # TODO: Do not use envvar! This has to be provided by the user somehow
-        android_ndk = self._conanfile.conf["tools.android:ndk_path"]
-        if not android_ndk:
+        android_ndk_path = self._conanfile.conf["tools.android:ndk_path"]
+        if not android_ndk_path:
             raise ConanException('CMakeToolchain needs tools.android:ndk_path configuration defined')
-        android_ndk = android_ndk.replace("\\", "/")
+        android_ndk_path = android_ndk_path.replace("\\", "/")
 
         ctxt_toolchain = {
-            'CMAKE_SYSTEM_VERSION': self._conanfile.settings.os.api_level,
-            'CMAKE_ANDROID_ARCH_ABI': android_abi,
-            'CMAKE_ANDROID_STL_TYPE': libcxx_str,
-            'CMAKE_ANDROID_NDK': android_ndk,
+            'ANDROID_PLATFORM': self._conanfile.settings.os.api_level,
+            'ANDROID_ABI': android_abi,
+            'ANDROID_STL': libcxx_str,
+            'ANDROID_NDK_PATH': android_ndk_path,
         }
         return ctxt_toolchain
 

